spirv_cross.cpp

```diff
@@ -3910,10 +3928,40 @@ void Compiler::CombinedImageSamplerUsageHandler::add_hierarchy_to_comparison_ids
 {
 	// Traverse the variable dependency hierarchy and tag everything in its path with comparison ids.
 	comparison_ids.insert(id);
+	
+	// If the underlying resource has been used for comparison then duplicate loads of that resource must be too.
+	if (!compiler.supports_combined_samplers())
+	{
+		for (auto it = dependency_hierarchy.begin(); it != dependency_hierarchy.end(); ++it)
+		{
+			if (it->second.find(id) != it->second.end())
+				comparison_ids.insert(it->first);
+		}
+	}
+	
 	for (auto &dep_id : dependency_hierarchy[id])
 		add_hierarchy_to_comparison_ids(dep_id);
 }
 
+// If the underlying resource has been used for comparison then duplicate loads of that resource must be too.
+bool Compiler::CombinedImageSamplerUsageHandler::dependent_used_for_comparison(uint32_t id) const
+{
+	if (compiler.supports_combined_samplers())
```

Comment:
Similar here to samplers. Is there a concrete test case for this in the PR?

Reply:
The generated Metal code of the test shader `shaders-msl/asm/frag/depth-compare.asm.frag` will fail to compile without this change.
A `depthcube<float>` texture will be sampled with `sample_compare` and also with `sample`. For both sampling functions the return value is `float` and not `float4` because its a depth texture even though it's sampled with a non-comparison sample function in the second case.
This case (and the above section) fixes that issue. Otherwise a `.x` suffix is added, which is not allowed for floating point types.

Reply:
That should already work, but I guess you found some edge cases. I'll check it out in more detail.

Reply:
Hm, I see no difference running master and this PR on that file both with and without optimization. I'll try to create an isolated reproducer, because I don't really understand this particular change.

Reply:
Ok, have a reproducer which this PR partially fixes:
```
#version 450

layout(set = 0, binding = 0) uniform texture2D uTexture;
layout(set = 0, binding = 1) uniform sampler uSampler;
layout(set = 0, binding = 2) uniform samplerShadow uSamplerShadow;

layout(location = 0) out float FragColor;
layout(location = 0) in vec3 vUV;

float sample_normal2(texture2D tex)
{
	return texture(sampler2D(tex, uSampler), vUV.xy).x;
}

float sample_normal(texture2D tex)
{
	return sample_normal2(tex);
}

float sample_comp(texture2D tex)
{
	return texture(sampler2DShadow(tex, uSamplerShadow), vUV);
}

void main()
{
	FragColor = sample_normal(uTexture);
	FragColor += sample_comp(uTexture);
}
```
The generated code becomes:
```
#pragma clang diagnostic ignored "-Wmissing-prototypes"

#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct main0_out
{
    float FragColor [[color(0)]];
};

struct main0_in
{
    float3 vUV [[user(locn0)]];
};

static inline __attribute__((always_inline))
float sample_normal2(thread const texture2d<float> tex, thread sampler uSampler, thread float3& vUV)
{
    return tex.sample(uSampler, vUV.xy).x;
}

static inline __attribute__((always_inline))
float sample_normal(thread const depth2d<float> tex, thread sampler uSampler, thread float3& vUV)
{
    return sample_normal2(tex, uSampler, vUV);
}

static inline __attribute__((always_inline))
float sample_comp(thread const depth2d<float> tex, thread float3& vUV, thread sampler uSamplerShadow)
{
    return tex.sample_compare(uSamplerShadow, vUV.xy, vUV.z);
}

fragment main0_out main0(main0_in in [[stage_in]], depth2d<float> uTexture [[texture(0)]], sampler uSampler [[sampler(0)]], sampler uSamplerShadow [[sampler(1)]])
{
    main0_out out = {};
    out.FragColor = sample_normal(uTexture, uSampler, in.vUV);
    out.FragColor += sample_comp(uTexture, in.vUV, uSamplerShadow);
    return out;
}
```
The parameters have to end up as depth2d as well, but we only get one link of the hierarchy resolved. I'll try to think of something here ...

Reply:
Great, thanks for the effort. I'd be happy to see a simpler solution that doesn't have O(n^2) complexity as you pointed out.

Reply:
Now, I could implement this on master tomorrow and you can rebase against that, but if rebasing is too hard, I can do a trampoline branch where I merge your branch, add my fix on top and push that to master. Which method do you prefer?

Reply:
I'm sorry I missed your reply here. If you have a better solution I'd be happy to replace it. Either way works for me. Since you mentioned you might want to make a few adjustments to this PR yourself, a side branch might be a good way to handle it.